### PR TITLE
Feature: ros2topic echo filtering with python expression

### PR DESCRIPTION
**Pull request by matthews-jca**
Based on the ros1 topic echo filtering code, porting this feature over to ros2. Added an all pass and all reject test

### CODE REVIEW

**Comment by matthews-jca** [comment]
Basic implementation for https://github.com/ros2/ros2cli/issues/651

File: ros2topic/ros2topic/verb/echo.py [hunk 1]
```diff
@@ -112,6 +112,10 @@ def add_arguments(self, parser, cli_name):
             '--no-lost-messages', action='store_true', help="Don't report when a message is lost")
         parser.add_argument(
             '--raw', action='store_true', help='Echo the raw binary representation')
+        parser.add_argument(
+            '--filter', dest='filter_expr', help='Python expression to filter messages that '
+                                                 'are printed.Expression can use Python builtins '
```
Comment (fujitatomoya) [inline ros2topic/ros2topic/verb/echo.py:117]:
```suggestion
                                                 'are printed. Expression can use Python builtins '
```

File: ros2topic/test/test_echo_pub.py [hunk 2]
```diff
@@ -311,6 +311,57 @@ def publish_message():
...
+                            filtered_rmw_implementation=get_rmw_implementation_identifier()
+                        )
+                    ) as command:
+                        # The future won't complete - we will hit the timeout
+                        self.executor.spin_until_future_complete(
+                            rclpy.task.Future(), timeout_sec=5
+                        )
+                    command.wait_for_shutdown(timeout=10)
+                    # Check results
+
```
Comment (fujitatomoya) [inline ros2topic/test/test_echo_pub.py:354]: do we need this line?

**Review by fujitatomoya — COMMENTED** [review]
this can bring up the same user experience with `ros topic --filter`, looks good to me.

**Comment by fujitatomoya** [comment]
CC: @clalancette

File: ros2topic/test/test_echo_pub.py (hunk 2 above)
Comment (matthews-jca) [reply ros2topic/test/test_echo_pub.py:354]: I'm not quite as familiar with the launch testing, but the existing tests used this format as the echo verb never times out on its own so the inner check on spin until future complete is actually what kills the process. Launch wait for shutdown is just to make sure the actual ros processes are cleaned up, I believe

Comment (fujitatomoya) [reply ros2topic/test/test_echo_pub.py:354]: sorry i wasn't clear. it is just a nitpick, do we need extra line 354?

Comment (matthews-jca) [reply ros2topic/test/test_echo_pub.py:354]: Oh, not at all. If anything I'd flip 353 and 354. I haven't made many PRs to the ROS codebase yet, so I appreciate the nitpicks.

**Review by fujitatomoya — CHANGES_REQUESTED** [review]
Can you address DCO error?

**Comment by matthews-jca** [comment]
> 
> 
> Can you address DCO error?

The commit the DCO bot is showing as an error was while accepting your suggested change on https://github.com/ros2/ros2cli/pull/654/commits/d42327453fbe331f86344a518439fb06f11518b9

I tried following the bots suggestion to fix the sign off, but it doesn't appear to be taking to the older commit.

**Comment by matthews-jca** [comment]
I've tried everything I could think of to resolve the DCO issue on the co-authored changed I approved through the Github interface above on https://github.com/ros2/ros2cli/commit/d42327453fbe331f86344a518439fb06f11518b9 but it is still complaining about it. Can anyone point me in the right direction on how to resolve this?

**Comment by hidmic** [comment]
@matthews-jca have you attempted to rebase and amend the commit message of https://github.com/ros2/ros2cli/commit/d42327453fbe331f86344a518439fb06f11518b9?

**Comment by matthews-jca** [comment]
> 
> 
> @matthews-jca have you attempted to rebase and amend the commit message of [d423274](https://github.com/ros2/ros2cli/commit/d42327453fbe331f86344a518439fb06f11518b9)?

Correct. I've tried a couple times based on the instructions on how to fix DCO errors and been unable to get the check to pass

**Comment by hidmic** [comment]
> Correct. I've tried a couple times based on the instructions on how to fix DCO errors and been unable to get the check to pass.

Hmm, just to be clear, you did amend the offending commit during an interactive rebase against `master` and that failed? Our DCO check just looks for a sign-off line in the commit message. I wonder why simply editing the commit message didn't work.

**Comment by matthews-jca** [comment]
> 
> Hmm, just to be clear, you did amend the offending commit during an interactive rebase against `master` and that failed? Our DCO check just looks for a sign-off line in the commit message. I wonder why simply editing the commit message didn't work.

I've tried both the instructions from the failed check details (https://github.com/ros2/ros2cli/pull/654/checks?check_run_id=3181176822) as well as going back to my branch and forcing an amended commit message including a sign off on the co-authored changed, but nothing seems to be making it happy

**Review by audrow — APPROVED** [review]
I've squashed and rebased this PR. The approach and test look good to me.

**Comment by audrow** [comment]
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/15188/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9908/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12868/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15367/)

**Comment by audrow** [comment]
Thank you for the PR @matthews-jca!

**Pull request merged**